

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>MMMx|atomic &mdash; MMMx  documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="/source/_static/custom.css" type="text/css" />
  <link rel="stylesheet" href="/source/_static/theme_override.css" type="text/css" />
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  <link rel="stylesheet" href="_static/table_util.css" type="text/css" />
  <link rel="stylesheet" href="_static/table_main.css" type="text/css" />
  <link rel="stylesheet" href="_static/theme_override.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="MMMx|RigiFlex" href="MMMx_RigiFlex.html" />
    <link rel="prev" title="Selection by address" href="MMMx_addresses.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Ensemble structure</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="functional_entity.html">Functional entity</a><ul>
<li class="toctree-l2"><a class="reference internal" href="functional_entity.html#the-object-of-interest">The object of interest</a></li>
<li class="toctree-l2"><a class="reference internal" href="functional_entity.html#structural-hierarchy">Structural hierarchy</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="structure_representation.html">Structure representation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="structure_representation.html#protein-database-pdb-format">Protein database (PDB) format</a></li>
<li class="toctree-l2"><a class="reference internal" href="structure_representation.html#pdbx-mmcif-format">PDBx/mmCIF format</a></li>
<li class="toctree-l2"><a class="reference internal" href="structure_representation.html#mmmx-internal-representations">MMMx internal representations</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="MMMx_addresses.html">Selection by address</a><ul>
<li class="toctree-l2"><a class="reference internal" href="MMMx_addresses.html#topology-and-ensemble-addressing">Topology and ensemble addressing</a></li>
<li class="toctree-l2"><a class="reference internal" href="MMMx_addresses.html#chains">Chains</a></li>
<li class="toctree-l2"><a class="reference internal" href="MMMx_addresses.html#residues">Residues</a></li>
<li class="toctree-l2"><a class="reference internal" href="MMMx_addresses.html#atoms">Atoms</a></li>
<li class="toctree-l2"><a class="reference internal" href="MMMx_addresses.html#water-molecules">Water molecules</a></li>
<li class="toctree-l2"><a class="reference internal" href="MMMx_addresses.html#conformers">Conformers</a></li>
<li class="toctree-l2"><a class="reference internal" href="MMMx_addresses.html#rotamers">Rotamers</a></li>
<li class="toctree-l2"><a class="reference internal" href="MMMx_addresses.html#locations">Locations</a></li>
<li class="toctree-l2"><a class="reference internal" href="MMMx_addresses.html#programmatic-access">Programmatic access</a></li>
<li class="toctree-l2"><a class="reference internal" href="MMMx_addresses.html#changes-compared-to-mmm">Changes compared to MMM</a></li>
<li class="toctree-l2"><a class="reference internal" href="MMMx_addresses.html#correspondence-with-chimerax">Correspondence with ChimeraX</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">MMMx|atomic</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#basic-considerations">Basic considerations</a></li>
<li class="toctree-l2"><a class="reference internal" href="#topology">Topology</a></li>
<li class="toctree-l2"><a class="reference internal" href="#conformation">Conformation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#element-assignment-of-atoms">Element assignment of atoms</a></li>
<li class="toctree-l2"><a class="reference internal" href="#conversion-to-pdb-representation">Conversion to PDB representation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#conversion-from-pdb-representation">Conversion from PDB representation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="MMMx_RigiFlex.html">MMMx|RigiFlex</a><ul>
<li class="toctree-l2"><a class="reference internal" href="MMMx_RigiFlex.html#basic-considerations">Basic considerations</a></li>
<li class="toctree-l2"><a class="reference internal" href="MMMx_RigiFlex.html#format">Format</a></li>
<li class="toctree-l2"><a class="reference internal" href="MMMx_RigiFlex.html#conversion-to-mmmx-atomic-representation">Conversion to MMMx|atomic representation</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Spectroscopic labels</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="rotamer_concept.html">Rotamer libraries</a><ul>
<li class="toctree-l2"><a class="reference internal" href="rotamer_concept.html#concept">Concept</a></li>
<li class="toctree-l2"><a class="reference internal" href="rotamer_concept.html#implementation">Implementation</a></li>
<li class="toctree-l2"><a class="reference internal" href="rotamer_concept.html#the-get-label-function">The <code class="docutils literal notranslate"><span class="pre">get_label</span></code> function</a></li>
<li class="toctree-l2"><a class="reference internal" href="rotamer_concept.html#set-of-libraries">Set of libraries</a></li>
<li class="toctree-l2"><a class="reference internal" href="rotamer_concept.html#rotamer-library-format">Rotamer library format</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="distance_distributions.html">Distance distributions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="distance_distributions.html#concept">Concept</a></li>
<li class="toctree-l2"><a class="reference internal" href="distance_distributions.html#generic-syntax">Generic syntax</a></li>
<li class="toctree-l2"><a class="reference internal" href="distance_distributions.html#residue">Residue</a></li>
<li class="toctree-l2"><a class="reference internal" href="distance_distributions.html#atoms">Atoms</a></li>
<li class="toctree-l2"><a class="reference internal" href="distance_distributions.html#locations">Locations</a></li>
<li class="toctree-l2"><a class="reference internal" href="distance_distributions.html#coordinates-any-level">Coordinates (any level)</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Programmer's Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="object_access.html">Get object attributes</a><ul>
<li class="toctree-l2"><a class="reference internal" href="object_access.html#concept">Concept</a></li>
<li class="toctree-l2"><a class="reference internal" href="object_access.html#generic-syntax">Generic syntax</a></li>
<li class="toctree-l2"><a class="reference internal" href="object_access.html#chain">Chain</a></li>
<li class="toctree-l2"><a class="reference internal" href="object_access.html#residue">Residue</a></li>
<li class="toctree-l2"><a class="reference internal" href="object_access.html#atoms">Atoms</a></li>
<li class="toctree-l2"><a class="reference internal" href="object_access.html#locations">Locations</a></li>
<li class="toctree-l2"><a class="reference internal" href="object_access.html#single-conformer">Single conformer</a></li>
<li class="toctree-l2"><a class="reference internal" href="object_access.html#coordinates-atomic-numbers-any-level">Coordinates &amp; atomic numbers (any level)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="object_modification.html">Set object attributes</a><ul>
<li class="toctree-l2"><a class="reference internal" href="object_modification.html#concept">Concept</a></li>
<li class="toctree-l2"><a class="reference internal" href="object_modification.html#generic-syntax">Generic syntax</a></li>
<li class="toctree-l2"><a class="reference internal" href="object_modification.html#residue">Residue</a></li>
<li class="toctree-l2"><a class="reference internal" href="object_modification.html#atoms">Atoms</a></li>
<li class="toctree-l2"><a class="reference internal" href="object_modification.html#locations">Locations</a></li>
<li class="toctree-l2"><a class="reference internal" href="object_modification.html#coordinates-any-level">Coordinates (any level)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="property_functions.html">Property functions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="property_functions.html#mass">Mass</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="ChimeraX.html">ChimeraX</a><ul>
<li class="toctree-l2"><a class="reference internal" href="ChimeraX.html#remarks">Remarks</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="python.html">Python</a><ul>
<li class="toctree-l2"><a class="reference internal" href="python.html#remarks">Remarks</a></li>
<li class="toctree-l2"><a class="reference internal" href="python.html#current-conclusion">Current conclusion</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="dependencies.html">Dependencies</a></li>
</ul>
<p class="caption"><span class="caption-text">Architecture</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="concept.html">Concept</a><ul>
<li class="toctree-l2"><a class="reference internal" href="concept.html#open-source-concept">Open source concept</a></li>
<li class="toctree-l2"><a class="reference internal" href="concept.html#separation-of-user-interface-from-computation">Separation of user interface from computation</a></li>
<li class="toctree-l2"><a class="reference internal" href="concept.html#visibility-to-users-and-programmers">Visibility to users and programmers</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="data.html">Data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="data.html#entity">Entity</a></li>
<li class="toctree-l2"><a class="reference internal" href="data.html#rotamer-library">Rotamer library</a></li>
<li class="toctree-l2"><a class="reference internal" href="data.html#distance-distribution-restraints">Distance distribution restraints</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="docs.html">Documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="docs.html#concept">Concept</a></li>
<li class="toctree-l2"><a class="reference internal" href="docs.html#automation">Automation</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MMMx</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>MMMx|atomic</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="mmmx-atomic">
<span id="id1"></span><h1>MMMx|atomic<a class="headerlink" href="#mmmx-atomic" title="Permalink to this headline">¶</a></h1>
<div class="section" id="basic-considerations">
<h2>Basic considerations<a class="headerlink" href="#basic-considerations" title="Permalink to this headline">¶</a></h2>
<p>From a thermodynamic point of view, different conformers have different probability to occur, as they have different free energies. They should thus be assigned populations.</p>
<p>If modelling is based on restraints in the form of probability density distributions, such as distance distribution restraints, probabilities of conformers are accessible in modelling.</p>
<p>The reason for alternate atom locations in protein crystal structures is, in most cases, significant population of more than one sidechain rotamer.
This should be assigned at residue level, not atom level.</p>
<p>From both a conceptual and data processing point of view, it makes sense to separate topology (fixed and the same for all conformers) from coordinates (variable).</p>
<p>Coordinates are heterogeneously different, meaning that for some atoms they do differ between conformers, while for others they do not.
Especially for large ensembles, it makes sense to store only unique coordinates.</p>
<div class="admonition-coordinate-indices admonition">
<p class="admonition-title">Coordinate indices</p>
<p>An atom coordinate is assigned by five indices. Three are <em>topological</em> and designate</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>the molecule, which can be a macromolecular chain</p></li>
<li><p>the residue in a chain</p></li>
<li><p>the atom in a residue</p></li>
</ol>
<p>The remaining two indices are <em>conformational</em> and designate</p>
<ol class="arabic simple" start="4">
<li><p>the conformer</p></li>
<li><p>the rotamer of a sidegroup (structures with specified rotamers) or one of the atom locations existing for a residue (crystal structures with locations)</p></li>
</ol>
</div></blockquote>
</div>
<p>The Cartesian coordinate array <img class="math" src="_images/math/91bb7ebe5360f220c5ddca9f8a3cff6379617fe0.svg" alt="\mathbf{C}"/> of an entity is thus supplemented by a coordinate index array <img class="math" src="_images/math/3e96f1d3536f5bc432fbb8d17a3b199f92aacff3.svg" alt="\mathbf{I}_\mathrm{c}"/>.
In Matlab, <a class="reference external" href="https://blogs.mathworks.com/loren/2013/02/20/logical-indexing-multiple-conditions/">logical indexing</a>, together with the index array, provides a very intuitive and efficient way for extracting required subsets of coordinates and to reassign them after some processing.
The index array is typed <code class="docutils literal notranslate"><span class="pre">uint16</span></code>, so that each hierarchy level can hold up to 65535 items.</p>
<p>The coordinate array is further supplemented by a vector of atomic numbers (number of the element in the periodic table) and by a vector of occupancies. Both vectors are single-byte unsigned integers. Occupancy is multiplied by 100 and rounded to an integer for storage.</p>
</div>
<div class="section" id="topology">
<h2>Topology<a class="headerlink" href="#topology" title="Permalink to this headline">¶</a></h2>
<p>Topology describes chemical structure and is immutable for an entity. In other words, any chemical reaction, such as posttranslational modification, is a transfer from one entity to another entity.</p>
<p>Topology is represented in terms of the number of objects (chains, molecules, residues, atoms), their identifiers, and their hierarchical relation.
Hierarchical relation means that an atom belongs to a residue or molecule and a residue belongs to a chain. The term <em>residue</em> includes amino acid residues, nucleotide residues, and, if someone finds a way to characterize carbohydrates, a sugar residue.</p>
<p>In Matlab, hierarchy is easily represented by a multi-level structure. By combining this with <a class="reference external" href="https://ch.mathworks.com/help/matlab/matlab_prog/generate-field-names-from-variables.html">dynamic fieldnames</a>, storage and access become very convenient and efficient.</p>
<p>Further attributes of objects can be stored by extending levels of the topology variable with additional fields.</p>
<div class="admonition-field-name-conventions admonition">
<p class="admonition-title">Field name conventions</p>
<p>All object names start with capital letters. They are:</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>Chain identifiers</strong>. Small letters from PDB files with more than 26 chains are represented by a capital letter with an appended underscore.
Entities with more than 26 chains can be exchanged with other software only in <code class="docutils literal notranslate"><span class="pre">mmCIF</span></code> format.</p></li>
<li><p><strong>Residue identifiers</strong>. Capital <code class="docutils literal notranslate"><span class="pre">R</span></code> followed by the residue number.</p></li>
<li><p><strong>Atom identifiers</strong>. The atom name in PDB or mmCIF syntax. Primes are substituted by an underscore.
If a name starting with a lowercase character is encountered, the first letter is capitalized and an underscore is appended to the atom name.</p></li>
</ul>
<p>The fieldname <code class="docutils literal notranslate"><span class="pre">selected</span></code> is reserved at all levels. At entity level it is a vector of selected conformers (default: 1), at other levels a boolean flag.</p>
<p>The fieldname <code class="docutils literal notranslate"><span class="pre">index</span></code> is reserved at all levels below entity. This field holds the topology index per object and level.</p>
<p>The fieldnames <code class="docutils literal notranslate"><span class="pre">water</span></code> and <code class="docutils literal notranslate"><span class="pre">water_selected</span></code> at entity level are reserved for water molecules. They hold an index vector of all water atoms and a flag whether they are selected.</p>
<p>The fieldnames <code class="docutils literal notranslate"><span class="pre">name</span></code>, <code class="docutils literal notranslate"><span class="pre">xyz</span></code>, <code class="docutils literal notranslate"><span class="pre">index_array</span></code>, <code class="docutils literal notranslate"><span class="pre">elements</span></code>, <code class="docutils literal notranslate"><span class="pre">occupancies</span></code>, and <code class="docutils literal notranslate"><span class="pre">original_residue_numbers</span></code> are reserved at entity level</p>
<p>The fieldname <code class="docutils literal notranslate"><span class="pre">populations</span></code> is reserved at entity level and at residue level. At residue level, the length of the population vecor specifies the number of rotamers.</p>
<p>The fieldnames <code class="docutils literal notranslate"><span class="pre">`dssp</span></code>, <code class="docutils literal notranslate"><span class="pre">labels</span></code>, <code class="docutils literal notranslate"><span class="pre">locations</span></code>, <img class="math" src="_images/math/7c2336375db97939f3533428a65c091e8bb3c5c9.svg" alt="name"/>, <code class="docutils literal notranslate"><span class="pre">selected_rotamers</span></code> and <code class="docutils literal notranslate"><span class="pre">sheet</span></code> are reserved at residue level</p>
<p>The fieldnames <code class="docutils literal notranslate"><span class="pre">element</span></code>, <code class="docutils literal notranslate"><span class="pre">selected_locations</span></code>, <code class="docutils literal notranslate"><span class="pre">charge</span></code>, and <code class="docutils literal notranslate"><span class="pre">bfactor</span></code> are reserved at atom level</p>
<p>Methods can define further lower-case fieldnames for object attributes. These attributes are internal to the method.
Passing back such attributes to MMMx and saving them in <code class="docutils literal notranslate"><span class="pre">mmCIF</span></code> output files may be enabled in a future release.</p>
</div></blockquote>
</div>
</div>
<div class="section" id="conformation">
<h2>Conformation<a class="headerlink" href="#conformation" title="Permalink to this headline">¶</a></h2>
<p>The only information on conformation that the topology variable needs to hold are population vectors (vide supra) on entity level and on residue level (rotamers).
The length of these vectors specifies the number of entity conformers or sidechain rotamers. The topology variable must also hold the topological indices (vide supra).
With that information, all atom coordinates can be retrieved via the index array from the coordinate array and the associated populations can be computed.</p>
</div>
<div class="section" id="element-assignment-of-atoms">
<h2>Element assignment of atoms<a class="headerlink" href="#element-assignment-of-atoms" title="Permalink to this headline">¶</a></h2>
<p>The information still missing even for basic computations concerns element assignment of the atoms, as retrieval from the atom name is not safe.
For ease of access, this information is stored in a separate element vector that matches the coordinate index array, but can be typed <img class="math" src="_images/math/3e500699666decfd91c08603c6f92c088357f12f.svg" alt="int8"/>.</p>
</div>
<div class="section" id="conversion-to-pdb-representation">
<h2>Conversion to PDB representation<a class="headerlink" href="#conversion-to-pdb-representation" title="Permalink to this headline">¶</a></h2>
<p>In the <code class="docutils literal notranslate"><span class="pre">MMMx|atomic</span></code> representation, the same atom coordinate can apply in several conformers (<em>models</em> in the PDB representation).
Upon conversion to PDB, the coordinate array then expands. The PDB writer of MMMx indexes rather than expands during writing to reduce memory requirements.</p>
<p>Conformer populations are stored in the free-text section of a <code class="docutils literal notranslate"><span class="pre">REMARK</span> <span class="pre">400</span> <span class="pre">COMPOUND</span></code> record in the following format (example):</p>
<p><code class="docutils literal notranslate"><span class="pre">REMARK</span> <span class="pre">400</span>&#160;&#160; <span class="pre">MODEL</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">1</span> <span class="pre">POPULATION</span>&#160;&#160; <span class="pre">0.5000</span></code></p>
<p>where the conformer number is in columns 19-28 of the line and the population in columns 40-48.</p>
<p>Rotameric states are expressed by alternate atom locations. Up to 26 (preferably) or 52 (with lower-case location identifiers) rotamers can be converted.
Not all external programs may be able to process PDB files with more than 26 locations. By default, only the 26 rotamers with highest populations are converted.
As an option, 52 rotamers can be converted.</p>
<p>If an atom coordinate in the <code class="docutils literal notranslate"><span class="pre">MMMx|atomic</span></code> representation is “not a number”, this atom is ignored.
This should happen only if the structure originated from an inconsistent PDB file.</p>
<p>MMMx converts to PDB representation only for two purposes:</p>
<ul class="simple">
<li><p>saving structure in a PDB files</p></li>
<li><p>transmitting structure to ChimeraX for visualization</p></li>
</ul>
<p>For saving to PDB, use function <code class="docutils literal notranslate"><span class="pre">put_pdb</span></code></p>
<div class="highlight-matlab notranslate"><div class="highlight"><pre><span></span><span class="n">exceptions</span> <span class="p">=</span> <span class="n">put_pdb</span><span class="p">(</span><span class="n">entity</span><span class="p">)</span>
<span class="n">exceptions</span> <span class="p">=</span> <span class="n">put_pdb</span><span class="p">(</span><span class="n">entity</span><span class="p">,</span><span class="n">fname</span><span class="p">)</span>
<span class="n">exceptions</span> <span class="p">=</span> <span class="n">put_pdb</span><span class="p">(</span><span class="n">entity</span><span class="p">,</span><span class="n">fname</span><span class="p">,</span><span class="n">options</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Parameters</dt><dd><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">entity</span></code> - entity in MMMx:atomic format</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">fname</span></code> - output filename (string), ‘.pdb’ is appended, if no extension, defaults to <code class="docutils literal notranslate"><span class="pre">entity.name</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">options</span></code> - output control, see below</p></li>
</ul>
</dd>
<dt>Returns</dt><dd><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">argout</span></code> - output arguments (<em>M</em>-element cell array)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">exceptions</span></code> - Matlab MException object, if any error occurred (1-element cell array)</p></li>
</ul>
</dd>
</dl>
<p><strong>Options</strong></p>
<table class="docutils align-default">
<colgroup>
<col style="width: 18%" />
<col style="width: 38%" />
<col style="width: 26%" />
<col style="width: 18%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Variable</p></th>
<th class="head"><p>Explanation</p></th>
<th class="head"><p>Type</p></th>
<th class="head"><p>Default</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">.selected</span></code></p></td>
<td><p>if true, only selected objects are output</p></td>
<td><p>boolean</p></td>
<td><p>false</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">.chainIDs</span></code></p></td>
<td><p>rename <em>C</em> chains as compared to entity</p></td>
<td><p>cell string (C,2)</p></td>
<td><p>empty</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">.</span></code></p></td>
<td><p>.chainIDs(c,1): name in entity</p></td>
<td><p>string</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">.</span></code></p></td>
<td><p>.chainIDs(c,2): name in PDB output</p></td>
<td><p>string</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">.charged</span></code></p></td>
<td><p>if true, writes atom charges</p></td>
<td><p>boolean</p></td>
<td><p>false</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">.bfactor</span></code></p></td>
<td><p>if false, writes zero B factors</p></td>
<td><p>boolean</p></td>
<td><p>false</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">.order</span></code></p></td>
<td><p>changed conformer order, only listed conformers</p></td>
<td><p>(1,M) double</p></td>
<td><p>empty (all conformers)</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">.pop</span></code></p></td>
<td><p>write conformer populations (in REMARK 400)</p></td>
<td><p>boolean</p></td>
<td><p>false</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">.pdbid</span></code></p></td>
<td><p>(pseudo-)PDB identifier, 4 characters</p></td>
<td><p>string</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">entity.name(1:4)</span></code></p></td>
</tr>
</tbody>
</table>
<hr class="docutils" />
<p>Please note that <code class="docutils literal notranslate"><span class="pre">put_pdb</span></code> does not save water atoms. Specifying them makes sense in crystal structures, but not in modelled structures.</p>
</div>
<div class="section" id="conversion-from-pdb-representation">
<h2>Conversion from PDB representation<a class="headerlink" href="#conversion-from-pdb-representation" title="Permalink to this headline">¶</a></h2>
<p>Upon loading a PDB file, MMMx does not make an effort to preserve atom numbers and only a limited effort to preserve residue numbers of the original PDB entry.
Residue numbers are preserved in the about <a class="reference external" href="http://mmcif.wwpdb.org/dictionaries/mmcif_pdbx_v40.dic/Items/_atom_site.pdbx_PDB_ins_code.html">96.5% structures that do not use “insertion codes”</a> and only if they all are positive numbers and if, within the same chain, they appear in ascending order in the PDB file.
The entity has a field <code class="docutils literal notranslate"><span class="pre">original_residue_numbers</span></code> that indicates whether residue numbers were preserved.</p>
<p>The number of chain/molecule conformers equals the number of PDB <code class="docutils literal notranslate"><span class="pre">models</span></code> for all chains and molecules of the entity. Uniform populations are assumed.</p>
<p>The number of rotamers of a residue is as large as the maximum number of alternate locations among the atoms of this residue.
Rotamer populations are mean populations over all atoms which have this number of alternate locations.</p>
<p>In case of topological inconsistency between <em>models</em>, topology is determined by the first model encountered in the PDB file (regardless of its model number).
Only atom coordinates are read for further models. Surplus atom coordinates are ignored. Missing atom coordinates are assigned “not a number”.</p>
<p><code class="docutils literal notranslate"><span class="pre">get_pdb</span></code> reads conformer populations from <code class="docutils literal notranslate"><span class="pre">REMARK</span> <span class="pre">400</span></code> records in PDB files that were stored by MMMx. If other software needs to convey populations in PDB files,
the corresponding <code class="docutils literal notranslate"><span class="pre">REMARK</span> <span class="pre">400</span></code> lines must have the following format (example):</p>
<p><code class="docutils literal notranslate"><span class="pre">REMARK</span> <span class="pre">400</span>&#160;&#160; <span class="pre">MODEL</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">1</span> <span class="pre">POPULATION</span>&#160;&#160; <span class="pre">0.5000</span></code></p>
<p>where the conformer number is in columns 19-28 of the line and the population in columns 40-48.
The populations must be specified for all models in the coordinate section and must not be specified for models that do not appear in the coordinate section.
Otherwise, uniform populations are assigned, as if the <code class="docutils literal notranslate"><span class="pre">REMARK</span> <span class="pre">400</span></code> was missing.</p>
<p>For loading from PDB, use function <code class="docutils literal notranslate"><span class="pre">get_pdb</span></code></p>
<div class="highlight-matlab notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">entity</span><span class="p">,</span><span class="n">exceptions</span><span class="p">]</span> <span class="p">=</span> <span class="n">get_pdb</span><span class="p">(</span><span class="n">ident</span><span class="p">)</span>
<span class="p">[</span><span class="n">entity</span><span class="p">,</span><span class="n">exceptions</span><span class="p">]</span> <span class="p">=</span> <span class="n">get_pdb</span><span class="p">(</span><span class="n">ident</span><span class="p">,</span><span class="n">options</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Parameters</dt><dd><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">ident</span></code> - if 4 characters and no extension, interpreted as PDB identifier for download, otherwise as local file name</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">options</span></code> - input control, see below</p></li>
</ul>
</dd>
<dt>Returns</dt><dd><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">entity</span></code> - entity in MMMx:atomic format</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">exceptions</span></code> - cell array of Matlab MException object, if any errors occurred</p></li>
</ul>
</dd>
</dl>
<p><strong>Options</strong></p>
<table class="docutils align-default">
<colgroup>
<col style="width: 18%" />
<col style="width: 38%" />
<col style="width: 26%" />
<col style="width: 19%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Variable</p></th>
<th class="head"><p>Explanation</p></th>
<th class="head"><p>Type</p></th>
<th class="head"><p>Default</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">.dssp</span></code></p></td>
<td><p>if true, DSSP is assigned for first model</p></td>
<td><p>boolean</p></td>
<td><p>false</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">.name</span></code></p></td>
<td><p>name for the entity</p></td>
<td><p>string</p></td>
<td><p>PDB identifier/ <code class="docutils literal notranslate"><span class="pre">MMMX</span></code></p></td>
</tr>
</tbody>
</table>
<p>Note that DSSP is only assigned, if there are no insertion codes and all residue numbers are positive.
We recommend to retrieve DSSP information separately by calling <a class="reference internal" href="object_access.html#get-conformer"><span class="std std-ref">get_conformer</span></a> with attribute <code class="docutils literal notranslate"><span class="pre">dssp</span></code>.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="MMMx_RigiFlex.html" class="btn btn-neutral float-right" title="MMMx|RigiFlex" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="MMMx_addresses.html" class="btn btn-neutral float-left" title="Selection by address" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020-, Gunnar Jeschke

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(false);
      });
  </script>

  
  
    
   

</body>
</html>